README.md
![Active](https://www.repostatus.org/badges/latest/active.svg)
![Downloads](https://img.shields.io/npm/dt/cordova-plugin-wechat.svg)
![version](https://img.shields.io/npm/v/cordova-plugin-wechat/latest.svg)


# 重要说明

Due to the security upgrade of iOS 13 system version, the official WeChat SDK has been adapted from 1.8.6 with supports * Universal Links * mode jumps, and the validity check when sharing.


From version 3.0.0 of our plugin, we changed to the latest WeChat SDK. Before using it, you need to configure the *Universal Links* service, and pay attention to passing  the `universallink` variable when installing the plugin, otherwise it will not work properly.

If you don't want to use the new version features, you can fall back to version 3.0.0 prior

# cordova-plugin-wechat

A cordova plugin, a JS version of Wechat SDK


# Feature

* check wechat client is installed;
* Share text, image, link,music,video,miniprogram to wechat timeline,session or favorite;
* open wechat auth;
* send wechat payment request;
* open wechat miniprogram;

# Documentation

[文档](https://jasonz1987.github.io/cordova-wechat-docs/) · [教程](https://www.jason-z.com/course/3) · [FAQ](https://jasonz1987.github.io/cordova-wechat-docs/docs/faq) · [DEMO](https://jasonz1987.github.io/cordova-wechat-docs/docs/demo)

# Install

```shell
cordova plugin add cordova-plugin-wechat  --variable wechatappid=YOUR_WECHAT_APPID --variable universallink=YOUR_UNIVERSAL_LINK
```

```shell
cordova build ios
cordova build android
```

# Usage 

## Check if wechat is installed

```Javascript
Wechat.isInstalled(function (installed) {
    alert("Wechat installed: " + (installed ? "Yes" : "No"));
}, function (reason) {
    alert("Failed: " + reason);
});
```

## Authenticate using Wechat
```Javascript
var scope = "snsapi_userinfo",
    state = "_" + (+new Date());
Wechat.auth(scope, state, function (response) {
    // you may use response.code to get the access token.
    alert(JSON.stringify(response));
}, function (reason) {
    alert("Failed: " + reason);
});
```

## Share text
```Javascript
Wechat.share({
    text: "This is just a plain string",
    scene: Wechat.Scene.TIMELINE   // share to Timeline
}, function () {
    alert("Success");
}, function (reason) {
    alert("Failed: " + reason);
});
```

## Share media(e.g. link, photo, music, video etc)
```Javascript
Wechat.share({
    message: {
        title: "Hi, there",
        description: "This is description.",
        thumb: "www/img/thumbnail.png",
        mediaTagName: "TEST-TAG-001",
        messageExt: "这是第三方带的测试字段",
        messageAction: "<action>dotalist</action>",
        media: "YOUR_MEDIA_OBJECT_HERE"
    },
    scene: Wechat.Scene.TIMELINE   // share to Timeline
}, function () {
    alert("Success");
}, function (reason) {
    alert("Failed: " + reason);
});
```

### Share link
```Javascript
Wechat.share({
    message: {
        ...
        media: {
            type: Wechat.Type.WEBPAGE,
            webpageUrl: "http://www.jason-z.com"
        }
    },
    scene: Wechat.Scene.TIMELINE   // share to Timeline
}, function () {
    alert("Success");
}, function (reason) {
    alert("Failed: " + reason);
});
```

### Share mini program
```Javascript
Wechat.share({
    message: {
        ...
        media: {
            type: Wechat.Type.MINI,
            webpageUrl: "https://www.jason-z.com", // 兼容低版本的网页链接
            userName: "wxxxxxxxx", // 小程序原始id
            path: "user/info", // 小程序的页面路径
            hdImageData: "http://wwww.xxx.com/xx.jpg", // 程序新版本的预览图二进制数据 不超过128kb 支持 地址 base64 temp
            withShareTicket: true, // 是否使用带shareTicket的分享
            miniprogramType: Wechat.Mini.RELEASE 
        }
    },
    scene: Wechat.Scene.SESSION   // 小程序仅支持聊天界面
}, function () {
    alert("Success");
}, function (reason) {
    alert("Failed: " + reason);
});
```

## Send payment request
```Javascript
// See https://github.com/xu-li/cordova-plugin-wechat-example/blob/master/server/payment_demo.php for php demo
var params = {
    partnerid: '10000100', // merchant id
    prepayid: 'wx201411101639507cbf6ffd8b0779950874', // prepay id
    noncestr: '1add1a30ac87aa2db72f57a2375d8fec', // nonce
    timestamp: '1439531364', // timestamp
    sign: '0CB01533B8C1EF103065174F50BCA001', // signed string
};

Wechat.sendPaymentRequest(params, function () {
    alert("Success");
}, function (reason) {
    alert("Failed: " + reason);
});
```

## Choose invoices from card list
```Javascript
//offical doc https://mp.weixin.qq.com/wiki?t=resource/res_main&id=mp1496561749_f7T6D
var params = {
    timeStamp: '1510198391', // timeStamp
    signType: 'SHA1', // sign type
    cardSign: 'dff450eeeed08120159d285e79737173aec3df94', // cardSign
    nonceStr: '5598190f-5fb3-4bff-8314-fd189ab4e4b8', // nonce
};

Wechat.chooseInvoiceFromWX(params,function(data){
    console.log(data);
},function(){
    alert('error');
})
```

## open wechat mini program 
```Javascript
//offical doc https://open.weixin.qq.com/cgi-bin/showdocument?action=dir_list&t=resource/res_list&verify=1&id=21526646437Y6nEC&token=&lang=zh_CN
var params = {
    userName: 'gh_d43f693ca31f', // userName
    path: 'pages/index/index?name1=key1&name2=key2', // open mini program page
    miniprogramType: Wechat.Mini.RELEASE // Developer version, trial version, and official version are available for selection
};

Wechat.openMiniProgram(params,function(data){
    console.log(data); // data:{extMsg:""}  extMsg: Corresponds to the app-parameter attribute in the Mini Program component <button open-type="launchApp">
},function(){
    alert('error');
})
```

more usage  please see [https://jasonz1987.github.io/cordova-wechat-docs/docs/usages](https://jasonz1987.github.io/cordova-wechat-docs/docs/usages)

# Join us

QQ群：190808518 
[![cordova-wechat官方交流群](https://pub.idqqimg.com/wpa/images/group.png)](http://shang.qq.com/wpa/qunwpa?idkey=8279476de172cacb72a51a5630744316c0069620ad8b33be3abee243af2cc001)

# Donate

we need your support to improve open source software ,if we induce your develop time ,welcome to donate us.

[![paypal](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/jasonz1987/6.66)

![donate.png](donate.png)

# LICENSE

[MIT LICENSE](http://opensource.org/licenses/MIT)


CHANGELOG.md
# Change Log

## 3.0.0 - 2020-02-20
### Added
- IOS Opensdk 1.8.6.2
- Support *Universal Links*
### Fixed
- Remove IOS Deprecated Method

## 2.9.0 - 2019-08-23
### Fixed
- remove appid param from ios wechat payment

## 2.8.0 - 2019-04-26
### Added
- cordova 9.0 support
- cordova 7.0 support renew
- add share image permission request on android
- add share file path support

## 2.7.0 - 2019-03-02
### Added
- add share to wechat mini program
- add App opens WeChat mini program
### Fixed
- removed libstd c++ from xcode10
- fix android cordova version 7.0.0 install bug
- fix ios payment appid param

## 2.6.0 - 2019-01-05
### Fixed
- Fixed cordova android 7.0 install script bug

## 2.5.0 - 2018-12-26
### Fixed
- Android preferences code problem

## 2.4.0 - 2018-07-27
### Added
- Android compress thumb 

## 2.3.0 - 2018-04-27
### Removed
- Share with mini-program (See #427, #428)

## 2.2.0 - 2018-04-26
### Added
- Share with mini-program (pull request #417 by xingducai)
- Add sub merchant support (pull request #401 by wxp-mjs)

## 2.1.0 - 2017-11-16
### Added
- iOS OpenSDK1.8.0
- Choose invoices from card list

## 2.0.0 - 2017-07-01
### Changed
- Use android sdk from JCenter
- Upgrade cordova versions
- More improvements to android version of the plugin

## 1.4.0 - 2017-01-10
### Added
- iOS OpenSDK1.7.5
- Android libammsdk 3.1.1

## 1.3.0 - 2016-09-13
### Added
- iOS openSDK1.7.3

## 1.2.0 - 2016-07-13
### Added
- iOS openSDK1.7.2

### Changed
- Allow wechat auth on the device without wechat installed.

### Removed
- WeChatSDK_1.5

## 1.1.5 - 2016-06-13
### Fixed
- fix android src files not being copied during platform add hook

## 1.1.4 - 2016-05-11
### Fixed
- wxapi directory gets deleted after removing other plugins

## 1.1.3 - 2016-01-28
### Fixed
- Fixed package name with hyphen issue in android

## 1.1.2 - 2016-01-22
### Fixed
- Fixed sharing emotion issue (remote file) in Android

## 1.1.1 - 2016-01-18
### Fixed
- Fixed sharing emotion issue in Android

## 1.1.0 - 2016-01-15
### Added
- Support Base64 image

### Changed
- Rename payment parameters

### Fixed
- Fixed thumbnail image size issue in iOS

## 1.0.0 - 2016-01-08
- Initial release


src/ios/CDVWechat.m
//
//  CDVWechat.m
//  cordova-plugin-wechat
//
//  Created by xu.li on 12/23/13.
//
//

#import "CDVWechat.h"

static int const MAX_THUMBNAIL_SIZE = 320;

@implementation CDVWechat

#pragma mark "API"
- (void)pluginInitialize {
    NSString* appId = [[self.commandDelegate settings] objectForKey:@"wechatappid"];
    
    NSString* universalLink = [[self.commandDelegate settings] objectForKey:@"universallink"];
    
    if (appId && ![appId isEqualToString:self.wechatAppId]) {
        self.wechatAppId = appId;
        [WXApi registerApp: appId universalLink: universalLink];
        
        NSLog(@"cordova-plugin-wechat has been initialized. Wechat SDK Version: %@. APP_ID: %@.", [WXApi getApiVersion], appId);
    }
}

- (void)isWXAppInstalled:(CDVInvokedUrlCommand *)command
{
    CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:[WXApi isWXAppInstalled]];

    [self.commandDelegate sendPluginResult:commandResult callbackId:command.callbackId];
}

- (void)share:(CDVInvokedUrlCommand *)command
{
    // if not installed
    if (![WXApi isWXAppInstalled])
    {
        [self failWithCallbackID:command.callbackId withMessage:@"未安装微信"];
        return ;
    }

    // check arguments
    NSDictionary *params = [command.arguments objectAtIndex:0];
    if (!params)
    {
        [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
        return ;
    }

    // save the callback id
    self.currentCallbackId = command.callbackId;

    SendMessageToWXReq* req = [[SendMessageToWXReq alloc] init];

    // check the scene
    if ([params objectForKey:@"scene"])
    {
        req.scene = (int)[[params objectForKey:@"scene"] integerValue];
    }
    else
    {
        req.scene = WXSceneTimeline;
    }

    // message or text?
    NSDictionary *message = [params objectForKey:@"message"];

    if (message)
    {
        req.bText = NO;

        // async
        [self.commandDelegate runInBackground:^{
            req.message = [self buildSharingMessage:message];
            
            [WXApi sendReq:(BaseReq *)req completion:^(BOOL success) {
                if (!success) {
                    [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
                                   self.currentCallbackId = nil;
                }
            }];
        }];
    }
    else
    {
        req.bText = YES;
        req.text = [params objectForKey:@"text"];

        [WXApi sendReq:(BaseReq *)req completion:^(BOOL success) {
            if (!success) {
                [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
                               self.currentCallbackId = nil;
            }
        }];
    }
}

- (void)sendAuthRequest:(CDVInvokedUrlCommand *)command
{

    SendAuthReq* req =[[SendAuthReq alloc] init];

    // scope
    if ([command.arguments count] > 0)
    {
        req.scope = [command.arguments objectAtIndex:0];
    }
    else
    {
        req.scope = @"snsapi_userinfo";
    }

    // state
    if ([command.arguments count] > 1)
    {
        req.state = [command.arguments objectAtIndex:1];
    }

    [WXApi sendAuthReq:req viewController:self.viewController delegate:self completion:^(BOOL success) {
        if(success) {
            self.currentCallbackId = command.callbackId;
        } else {
            [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
        }
    }];
}

- (void)sendPaymentRequest:(CDVInvokedUrlCommand *)command
{
    // check arguments
    NSDictionary *params = [command.arguments objectAtIndex:0];
    if (!params)
    {
        [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
        return ;
    }

    // check required parameters
    NSArray *requiredParams;
    if ([params objectForKey:@"mch_id"])
    {
        requiredParams = @[@"mch_id", @"prepay_id", @"timestamp", @"nonce", @"sign"];
    }
    else
    {
        requiredParams = @[@"partnerid", @"prepayid", @"timestamp", @"noncestr", @"sign"];
    }

    for (NSString *key in requiredParams)
    {
        if (![params objectForKey:key])
        {
            [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
            return ;
        }
    }

    PayReq *req = [[PayReq alloc] init];

    // NSString *appId = [params objectForKey:requiredParams[5]];
    // if (appId && ![appId isEqualToString:self.wechatAppId]) {
    //     self.wechatAppId = appId;
    //     [WXApi registerApp: appId];
    // }

    req.partnerId = [params objectForKey:requiredParams[0]];
    req.prepayId = [params objectForKey:requiredParams[1]];
    req.timeStamp = [[params objectForKey:requiredParams[2]] intValue];
    req.nonceStr = [params objectForKey:requiredParams[3]];
    req.package = @"Sign=WXPay";
    req.sign = [params objectForKey:requiredParams[4]];
    
    [WXApi sendReq:(BaseReq *)req completion:^(BOOL success) {
       if (success) {
           self.currentCallbackId = command.callbackId;
       } else {
           [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
       }
    }];
}

- (void)chooseInvoiceFromWX:(CDVInvokedUrlCommand *)command
{
    NSDictionary *params = [command.arguments objectAtIndex:0];
    WXChooseInvoiceReq *req = [[WXChooseInvoiceReq alloc] init];
    req.cardSign = [params objectForKey:@"cardSign"];
    req.timeStamp = [[params objectForKey:@"timeStamp"] intValue];
    req.appID = [params objectForKey:@"appId"];
    req.nonceStr = [params objectForKey:@"nonceStr"];
    req.signType = [params objectForKey:@"signType"];
    
    [WXApi sendReq:(BaseReq *)req completion:^(BOOL success) {
        if (success) {
            self.currentCallbackId = command.callbackId;
        } else {
            [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
        }
    }];
}

- (void)jumpToBizProfile:(CDVInvokedUrlCommand *)command
{
    // check arguments
//    NSDictionary *params = [command.arguments objectAtIndex:0];
//    if (!params)
//    {
//        [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
//        return ;
//    }
//
//    // check required parameters
//    NSArray *requiredParams;
//    requiredParams = @[@"type", @"info"];
//
//    for (NSString *key in requiredParams)
//    {
//        if (![params objectForKey:key])
//        {
//            [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
//            return ;
//        }
//    }
//    JumpToBizProfileReq *req = [JumpToBizProfileReq new];
//    NSString *bizType =  [params objectForKey:requiredParams[0]];
//
//    if ([bizType isEqualToString:@"Normal"]) {
//        req.profileType = WXBizProfileType_Normal;
//        req.username = [params objectForKey:requiredParams[1]];
//    } else {
//        req.profileType = WXBizProfileType_Device;
//        req.extMsg = [params objectForKey:requiredParams[1]];
//    }
//
//    if ([WXApi sendReq:req])
//    {
//        // save the callback id
//        self.currentCallbackId = command.callbackId;
//    }
//    else
//    {
//        [self failWithCallbackID:command.callbackId withMessage:@"发送请求失败"];
//    }
}

- (void)jumpToWechat:(CDVInvokedUrlCommand *)command
{
    // check arguments
    NSString *url = [command.arguments objectAtIndex:0];
    if (!url || ![url hasPrefix:@"weixin://"])
    {
        [self failWithCallbackID:command.callbackId withMessage:@"参数格式错误"];
        return ;
    }

    NSURL *formatUrl = [NSURL URLWithString:[url stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]]];
    if ([[UIApplication sharedApplication] canOpenURL:formatUrl]) {
        NSDictionary *options = @{UIApplicationOpenURLOptionUniversalLinksOnly : @YES};
        [[UIApplication sharedApplication] openURL:formatUrl options:options completionHandler:nil];
    } else{
        [self failWithCallbackID:command.callbackId withMessage:@"未安装微信或其他错误"];
    }
    return ;
}



#pragma mark "WXApiDelegate"

/**
 * Not implemented
 */
- (void)onReq:(BaseReq *)req
{
    NSLog(@"%@", req);
}

- (void)onResp:(WXLaunchMiniProgramResp *)resp
{
    BOOL success = NO;
    NSString *message = @"Unknown";
    NSDictionary *response = nil;

    switch (resp.errCode)
    {
        case WXSuccess:
            success = YES;
            break;

        case WXErrCodeCommon:
            message = @"普通错误";
            break;

        case WXErrCodeUserCancel:
            message = @"用户点击取消并返回";
            break;

        case WXErrCodeSentFail:
            message = @"发送失败";
            break;

        case WXErrCodeAuthDeny:
            message = @"授权失败";
            break;

        case WXErrCodeUnsupport:
            message = @"微信不支持";
            break;

        default:
            message = @"未知错误";
    }

    if (success)
    {
        if ([resp isKindOfClass:[SendAuthResp class]])
        {
            // fix issue that lang and country could be nil for iPhone 6 which caused crash.
            SendAuthResp* authResp = (SendAuthResp*)resp;
            response = @{
                         @"code": authResp.code != nil ? authResp.code : @"",
                         @"state": authResp.state != nil ? authResp.state : @"",
                         @"lang": authResp.lang != nil ? authResp.lang : @"",
                         @"country": authResp.country != nil ? authResp.country : @"",
                         };

            CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];

            [self.commandDelegate sendPluginResult:commandResult callbackId:self.currentCallbackId];
        }
        else if([resp isKindOfClass:[WXChooseInvoiceResp class]]){
                    WXChooseInvoiceResp* invoiceResp = (WXChooseInvoiceResp *)resp;

        //            response = @{
        //                         @"data":invoiceResp.cardAry
        //                         }
                    NSMutableArray *arrM = [[NSMutableArray alloc] init];
                    NSDictionary *mutableDic = nil;
                    for(WXInvoiceItem *invoiceItem in invoiceResp.cardAry){
                        mutableDic = @{
                                       @"cardId": invoiceItem.cardId,
                                       @"encryptCode": invoiceItem.encryptCode,
                                       };
                        [arrM addObject:mutableDic];
                    }
                    response = @{
                                 @"data": arrM
                                 };
                    NSLog(@"response======= %@", response);
                    CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
                    [self.commandDelegate sendPluginResult:commandResult callbackId:self.currentCallbackId];
                }
        else if ([resp isKindOfClass:[WXLaunchMiniProgramResp class]])
        {
            NSString *extMsg = resp.extMsg;
            response = @{
                         @"extMsg": extMsg
                         };
            CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
            [self.commandDelegate sendPluginResult:commandResult callbackId:self.currentCallbackId];
        }
        else
        {
            [self successWithCallbackID:self.currentCallbackId];
        }
    }
    else
    {
        [self failWithCallbackID:self.currentCallbackId withMessage:message];
    }

    [self pluginInitialize];
    self.currentCallbackId = nil;
}

#pragma mark "CDVPlugin Overrides"

- (void)handleOpenURL:(NSNotification *)notification
{
    NSURL* url = [notification object];

    if ([url isKindOfClass:[NSURL class]] && [url.scheme isEqualToString:self.wechatAppId])
    {
        [WXApi handleOpenURL:url delegate:self];
    }
}

#pragma mark "Private methods"

- (WXMediaMessage *)buildSharingMessage:(NSDictionary *)message
{
    WXMediaMessage *wxMediaMessage = [WXMediaMessage message];
    wxMediaMessage.title = [message objectForKey:@"title"];
    wxMediaMessage.description = [message objectForKey:@"description"];
    wxMediaMessage.mediaTagName = [message objectForKey:@"mediaTagName"];
    wxMediaMessage.messageExt = [message objectForKey:@"messageExt"];
    wxMediaMessage.messageAction = [message objectForKey:@"messageAction"];
    if ([message objectForKey:@"thumb"])
    {
        [wxMediaMessage setThumbImage:[self getUIImageFromURL:[message objectForKey:@"thumb"]]];
    }

    // media parameters
    id mediaObject = nil;
    WXMiniProgramObject *object;
    NSDictionary *media = [message objectForKey:@"media"];
    // check types
    NSInteger type = [[media objectForKey:@"type"] integerValue];
    switch (type)
    {
        case CDVWXSharingTypeApp:
            mediaObject = [WXAppExtendObject object];
            ((WXAppExtendObject*)mediaObject).extInfo = [media objectForKey:@"extInfo"];
            ((WXAppExtendObject*)mediaObject).url = [media objectForKey:@"url"];
            break;

        case CDVWXSharingTypeEmotion:
            mediaObject = [WXEmoticonObject object];
            ((WXEmoticonObject*)mediaObject).emoticonData = [self getNSDataFromURL:[media objectForKey:@"emotion"]];
            break;

        case CDVWXSharingTypeFile:
            mediaObject = [WXFileObject object];
            ((WXFileObject*)mediaObject).fileData = [self getNSDataFromURL:[media objectForKey:@"file"]];
            ((WXFileObject*)mediaObject).fileExtension = [media objectForKey:@"fileExtension"];
            break;

        case CDVWXSharingTypeImage:
            mediaObject = [WXImageObject object];
            ((WXImageObject*)mediaObject).imageData = [self getNSDataFromURL:[media objectForKey:@"image"]];
            break;

        case CDVWXSharingTypeMusic:
            mediaObject = [WXMusicObject object];
            ((WXMusicObject*)mediaObject).musicUrl = [media objectForKey:@"musicUrl"];
            ((WXMusicObject*)mediaObject).musicDataUrl = [media objectForKey:@"musicDataUrl"];
            break;

        case CDVWXSharingTypeVideo:
            mediaObject = [WXVideoObject object];
            ((WXVideoObject*)mediaObject).videoUrl = [media objectForKey:@"videoUrl"];
            break;
        case CDVWXSharingTypeMini:
            object = [WXMiniProgramObject object];
            object.webpageUrl = [media objectForKey:@"webpageUrl"];
            object.userName = [media objectForKey:@"userName"];
            object.path = [media objectForKey:@"path"]; // pages/inbox/inbox?name1=key1&name=key2
            object.hdImageData = [self getNSDataFromURL:[media objectForKey:@"hdImageData"]];
            object.withShareTicket = [[media objectForKey:@"withShareTicket"] boolValue];
            object.miniProgramType = (int)[[media objectForKey:@"miniProgramType"] integerValue];
            wxMediaMessage.mediaObject = object;
            return wxMediaMessage;
            
        case CDVWXSharingTypeWebPage:
        default:
            mediaObject = [WXWebpageObject object];
            ((WXWebpageObject *)mediaObject).webpageUrl = [media objectForKey:@"webpageUrl"];
    }

    wxMediaMessage.mediaObject = mediaObject;
    return wxMediaMessage;
}

- (NSData *)getNSDataFromURL:(NSString *)url
{
    NSData *data = nil;

    if ([url hasPrefix:@"http://"] || [url hasPrefix:@"https://"])
    {
        data = [NSData dataWithContentsOfURL:[NSURL URLWithString:url]];
    }
    else if ([url hasPrefix:@"data:image"])
    {
        // a base 64 string
        NSURL *base64URL = [NSURL URLWithString:url];
        data = [NSData dataWithContentsOfURL:base64URL];
    }
    else if ([url rangeOfString:@"temp:"].length != 0)
    {
        url =  [NSTemporaryDirectory() stringByAppendingPathComponent:[url componentsSeparatedByString:@"temp:"][1]];
        data = [NSData dataWithContentsOfFile:url];
    }
    else
    {
        // local file
        url = [[NSBundle mainBundle] pathForResource:[url stringByDeletingPathExtension] ofType:[url pathExtension]];
        data = [NSData dataWithContentsOfFile:url];
    }

    return data;
}

- (UIImage *)getUIImageFromURL:(NSString *)url
{
    NSData *data = [self getNSDataFromURL:url];
    UIImage *image = [UIImage imageWithData:data];

    if (image.size.width > MAX_THUMBNAIL_SIZE || image.size.height > MAX_THUMBNAIL_SIZE)
    {
        CGFloat width = 0;
        CGFloat height = 0;

        // calculate size
        if (image.size.width > image.size.height)
        {
            width = MAX_THUMBNAIL_SIZE;
            height = width * image.size.height / image.size.width;
        }
        else
        {
            height = MAX_THUMBNAIL_SIZE;
            width = height * image.size.width / image.size.height;
        }

        // scale it
        UIGraphicsBeginImageContext(CGSizeMake(width, height));
        [image drawInRect:CGRectMake(0, 0, width, height)];
        UIImage *scaled = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();

        return scaled;
    }

    return image;
}

- (void)successWithCallbackID:(NSString *)callbackID
{
    [self successWithCallbackID:callbackID withMessage:@"OK"];
}

- (void)successWithCallbackID:(NSString *)callbackID withMessage:(NSString *)message
{
    CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:message];
    [self.commandDelegate sendPluginResult:commandResult callbackId:callbackID];
}

- (void)failWithCallbackID:(NSString *)callbackID withError:(NSError *)error
{
    [self failWithCallbackID:callbackID withMessage:[error localizedDescription]];
}

- (void)failWithCallbackID:(NSString *)callbackID withMessage:(NSString *)message
{
    CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:message];
    [self.commandDelegate sendPluginResult:commandResult callbackId:callbackID];
}

-  (void)openMiniProgram:(CDVInvokedUrlCommand *)command
{
    NSDictionary *params = [command.arguments objectAtIndex:0];
    WXLaunchMiniProgramReq *launchMiniProgramReq = [WXLaunchMiniProgramReq object];
    launchMiniProgramReq.userName = [params objectForKey:@"userName"];  //拉起的小程序的username
    launchMiniProgramReq.path = [params objectForKey:@"path"];    //拉起小程序页面的可带参路径，不填默认拉起小程序首页
    launchMiniProgramReq.miniProgramType = (int)[[params objectForKey:@"miniprogramType"] integerValue]; //拉起小程序的类型
    [WXApi sendReq:launchMiniProgramReq completion:^(BOOL success) {
        if(success) {
             self.currentCallbackId = command.callbackId;
        } else {
            [self failWithCallbackID:command.callbackId withMessage:@"打开请求失败"];
        }
    }];
}

- (BOOL)handleUserActivity:(NSUserActivity *)userActivity {
   return [WXApi handleOpenUniversalLink:userActivity delegate:self];
}

- (BOOL)handleWechatOpenURL:(NSURL *)url  {
    return [WXApi handleOpenURL:url delegate:self];
}

@end
